Bug: #7820 paginator ignores dbal type conversions in identifiers

Ocramius opened the pull request:
Fixes #7820

Props to @lyska for giving me the opportunity to hunt this one down!

Ocramius committed:
Failing test case for #7820 - paginator doesn't use custom ID types

When using a `Doctrine\ORM\Tools\Pagination\Paginator` to iterate over a query that has entities with a custom DBAL type used in the identifier, then `$id->__toString()` is used implicitly by PDO, instead of being converted by the `Doctrine\DBAL\Types` system.

In order to reproduce this, you must have identifiers implementing `#__toString()` (to allow the `UnitOfWork` to hash them) and other accessors that are used by the custom DBAL type during DB/PHP conversions. If `#__toString()` and the DBAL type conversions are asymmetric, then the paginator will fail to find records.

Tricky situation, but this very much affects `ramsey/uuid-doctrine` and anyone relying on the `uuid_binary`.

Ocramius committed:
Fixed #7820 - convert identifiers for `WHERE IN(?)` queries before binding parameters

This patch introduces new internal API on the `ResultSetMapping` class, which is responsible
for finding the type of the single column identifier of a DQL query selection root.

lcobucci commented on lib/Doctrine/ORM/Query/ResultSetMapping.php:
```diff
@@ -583,5 +589,40 @@ public function addMetaResult($alias, $columnName, $fieldName, $isIdentifierColu
 
+    /**
+     * Retrieves the DBAL type name for the single identifier column of the root of a selection.
+     * Composite identifiers not supported!
```
Is there any test covering exceptions when trying to paginate entities with composite identifiers? Perhaps it might be important to add one, especially because we're strengthening this restriction.

Ocramius replied:
I'll add one test specific to `ResultSetMapping`

lcobucci commented on lib/Doctrine/ORM/Tools/Pagination/Paginator.php:
```diff
@@ -150,14 +155,21 @@ public function getIterator()
+            $whereInQuery = $this->cloneQuery($this->query);
+            $em           = $subQuery->getEntityManager();
+            $connection   = $em->getConnection();
+            $idType       = $this->getIdentifiersQueryScalarResultType($subQuery, $em);
```
Wouldn't we keep things a bit more contained if we perform this conversion in `WhereInWalker`? There's a lot going on there too but we already have the class metadata and it might be easier to detect the type?

lcobucci replied:
And I think we would also mitigate possible performance impact due to double query parsing (although query cache reduces that)

Ocramius replied:
> Wouldn't we keep things a bit more contained if we perform this conversion in `WhereInWalker`? 

Initially attempted this, but `WhereInWalker` does not have access to the DQL parameters, only to the DQL string/AST.

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH7820Test.php:
```diff
@@ -0,0 +1,158 @@
+    /**
+     * @var GH7820LineText
+     * @Id()
+     * @Column(type="Doctrine\Tests\ORM\Functional\Ticket\GH7820LineTextType")
```
Yeap, will need to re-visit this, since it didn't like it last time I tried it.

Ocramius replied:
Not applicable: seems like I hit an annotation bug in this case, since the class is ignored (not part of `use` statements, after all.

Won't fix for this patch though.

lcobucci replied:
That's quite unexpected! Well, nothing we can sort out now :smile:

Ocramius committed:
As per discussion with @lcobucci, it is better to keep dragons where
there be dragons, and this change does indeed rewrite the previous
approach by moving the responsibility of type conversion on a query
object from the `Paginator` to the `WhereInWalker`, which already
has access to class metadata for the root of the selection (and can
reliably detect the root of the selection too)

Ocramius commented:
Green! @lcobucci this should be much better now :+1:

lcobucci commented on lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:
```diff
@@ -2017,7 +2017,7 @@ public function isIdentifierUuid()
      *
      * @param string $fieldName
      *
-     * @return \Doctrine\DBAL\Types\Type|string|null
```
Is this change necessary/correct? Doesn't it affect `ClassMetadataInfo#getTypeOfColumn()` too?

Ocramius replied:
It is correct though, since this endpoint always returns a `string`, and the code below does indeed rely on a `string` being retrieved.

If a DBAL type was returned (which is never the case), a string cast would not be acceptable.

lcobucci commented on lib/Doctrine/ORM/Tools/Pagination/Paginator.php:
```diff
@@ -150,14 +151,16 @@ public function getIterator()
+            $ids          = array_map('current', $foundIdRows);
+
             $this->appendTreeWalker($whereInQuery, WhereInWalker::class);
             $whereInQuery->setHint(WhereInWalker::HINT_PAGINATOR_ID_COUNT, count($ids));
```
I know it's not really related to your fix, but shouldn't we keep the matched IDs as a hint too? Would increase the cohesion a bit more too.

lcobucci replied:
We might leave this for v3, though.

mpdude replied:
Just for the record, #8797 works in the other direction to replace the `count()` with a bool, since the information is not actually needed (at this time) and it improves cache efficiency.

lcobucci commented:
@Ocramius It's muuuuch better indeed. Great job there!

Ocramius committed:
#7820 restricted return type of `Doctrine\ORM\Mapping\ClassMetadataInfo#getTypeOfColumn()`

This method will always return `string|null`, so we can safely
remove DBAL types from its possible return types.

Ocramius committed:
#7820 documented `PersisterHelper#getTypeOfField()` array return type

Array values are `string`, and the array is a packed array.

Ocramius committed:
#7820 use `PersisterHelper` to figure out identifier types

This logic was pre-existing, but I forgot about it while writing
#7820, therefore it was re-implemented inside this unit of
code. Now we just use the `PersisterHelper`, which does all
the nice and shiny identifier type discovery operations we need.

lcobucci reviewed (approved):
Awesome!

ksowa commented:
@Ocramius It looks like this PR caused another bug. 

I use `ramsey/uuid` library. When trying to get collection of objects with UUID as a Primary Key, it fails just after  `Doctrine\ORM\Tools\Pagination\Paginator::getIterator()`, which pulls PKs as scalar:

```
$foundIdRows = $subQuery->getScalarResult();
// ...
$whereInQuery = $this->cloneQuery($this->query);
$ids = array_map('current', $foundIdRows);
```
In the `v2.6.3` the value was properly converting from the string to PHP value and then back to database value (`convertToPHPValue()` -> `convertToDatabaseValue()`). Since `v2.6.4` the `convertToDatabaseValue` is called first, which causes an error. 

I'm not sure whether this is issue in `ramsey/uuid` or Doctrine, but I think the change caused BC break.

Ocramius commented:
Note that I'm working specifically on this issue (and #7865) because of `ramsey/uuid`

akorz commented:
@Ocramius @ksowa the same issue for me
